### PR TITLE
Implement register + test

### PR DESCRIPTION
**Pull request by Roma1990Melnichuk**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationService.java [hunk 1]
```diff
@@ -4,4 +4,5 @@
 
 public interface RegistrationService {
     User register(User user);
+
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationService.java:7]: revert

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -5,10 +5,26 @@
...
+    private static final int MINIMUM_PASSWORD_LENGTH = 6;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new ValidationException("LOGIN " + user.getLogin() + "is used");
+        }
+        if (user.getPassword().length() < MINIMUM_PASSWORD_LENGTH) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:18]: Let's add checking password / login not null

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 3]
```diff
@@ -1,8 +1,109 @@
...
+import core.basesyntax.model.User;
+import core.basesyntax.service.RegistrationServiceImpl;
+import core.basesyntax.service.ValidationException;
+import java.util.Random;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
 public class HelloWorldTest {
+    private RegistrationServiceImpl registrationService;
+    private User user = new User();
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/HelloWorldTest.java:17]: Let's create private Random random = new Random(); and use it

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 4]
```diff
@@ -1,8 +1,109 @@
...
+    private String password = "admin" + new Random().nextInt(6, 99999);
+    private int userAge = new Random().nextInt(18, 99);
+
+    @BeforeEach
+    public void setUp() {
+        registrationService = new RegistrationServiceImpl();
+    }
+
+    @Test
+    public void check_correct_login() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/HelloWorldTest.java:28]: Please rename all methods according to the task description

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 5]
```diff
@@ -1,8 +1,109 @@
...
+
+    @Test
+    public void check_two_users_notequals() {
+        User user1 = new User();
+        user1.setLogin("user1");
+        user1.setPassword("password1");
+        user1.setAge(25);
+
+        User user2 = new User();
+        user2.setLogin("user2"); // Изменили login для второго пользователя
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/HelloWorldTest.java:73]: Please remove comment. Also, try to use English )

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 6]
```diff
@@ -1,8 +1,109 @@
...
+        user2.setPassword("password1");
+        user2.setAge(25);
+
+        assertNotEquals(user1, user2);
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/HelloWorldTest.java:77]: I'm not quite understand the purpose of the test method

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 7]
```diff
@@ -1,8 +1,109 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.model.User;
+import core.basesyntax.service.RegistrationServiceImpl;
+import core.basesyntax.service.ValidationException;
+import java.util.Random;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
 public class HelloWorldTest {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/HelloWorldTest.java:15]: Rename please the class name as you are testing `RegistrationService`, for example `RegistrationServiceTest`

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 8]
```diff
@@ -1,8 +1,109 @@
...
+        User user2 = new User();
+        user2.setLogin("user2"); // Изменили login для второго пользователя
+        user2.setPassword("password1");
+        user2.setAge(25);
+
+        assertNotEquals(user1, user2);
+    }
+
+    @Test
+    public void check_user_hashcode() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/HelloWorldTest.java:81]: Checking hashCode is pointless. 
When you are testing register() service method, you should check all test cases for this method: 

In your tests very important to check all possible edge cases. For example, we have min password length 6 characters. In this case would be good to test: empty password (0-length - not-ok), 'abc' (3-length - not-ok), 'abcdf' (5-length - not-ok edge case), 'abcdef' (6-length - ok edge case), 'abcdefgh' (8-length - ok). Few not-ok checks you can have in one not-ok test method. Same for ok-test. Testing only 'abc' (3-length - not-ok) and 'abcdefgh' (8-length - ok) is not enough because we do not check how the program behave with edge cases.

Make sure that after successful registration user was added to storage and register method returned correct user.

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 9]
```diff
@@ -5,10 +5,29 @@
...
+    private static final int MINIMUM_PASSWORD_LENGTH = 6;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new ValidationException("LOGIN " + user.getLogin() + "is used");
+        }
+        if (user.getPassword() != null) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:18]: If password is null we need to throw an exception to let client know about it

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 10]
```diff
@@ -5,10 +5,29 @@
...
+            if (user.getPassword().length() < MINIMUM_PASSWORD_LENGTH) {
+                throw new ValidationException("PASSWORD " + user.getPassword() + "is to short");
+            }
+        }
+
+        if (user.getLogin().length() < MINIMUM_LOGIN_LENGTH) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:24]: Add check login == null?

File: src/test/java/core/basesyntax/RegistrationServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        assertThrows(ValidationException.class, () -> registrationService.register(user));
+
+        failAge = -1;
+        user.setAge(failAge);
+        assertThrows(ValidationException.class, () -> registrationService.register(user));
+
+    }
+
+    @Test
+    public void accordance_twouserswithdifferentparametrs_notok() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/RegistrationServiceTest.java:109]: Pointless test method

File: src/test/java/core/basesyntax/RegistrationServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        user2.setLogin("");
+        user2.setPassword("");
+        user2.setAge(0);
+
+        assertThrows(ValidationException.class, () -> registrationService.register(user2));
+
+    }
+
+    @Test
+    public void register_setidafterauth_ok() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/RegistrationServiceTest.java:62]: There is no any logic about setting id in register method, so please remove this test case

File: src/test/java/core/basesyntax/RegistrationServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        assertThrows(ValidationException.class, () -> registrationService.register(user));
+
+        toShortPassword = null;
+        user.setPassword(toShortPassword);
+        assertThrows(NullPointerException.class, () -> registrationService.register(user));
+
+        toShortPassword = "trese";
+        user.setPassword(toShortPassword);
+        assertThrows(ValidationException.class, () -> registrationService.register(user));
+
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/RegistrationServiceTest.java:86]:
```suggestion
```
Check all places

File: src/test/java/core/basesyntax/RegistrationServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.model.User;
+import core.basesyntax.service.RegistrationServiceImpl;
+import core.basesyntax.service.ValidationException;
+import java.util.Random;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RegistrationServiceTest {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/RegistrationServiceTest.java:15]: Let's stick to some rules:
one test method - one test case, if you would like to check several parameters use ParameterizedTest
Tests should be isolated, don't need to check null, "" and min length in one test method. 
Also, name of test method should describe what you are going to test - register_validUser_ok, 
register_nullPassword_notOk etc

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 15]
```diff
@@ -5,10 +5,30 @@
...
     @Override
     public User register(User user) {
-        return null;
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new ValidationException("LOGIN " + user.getLogin() + "is used");
+        }
+
+        if (user.getPassword().length() < MINIMUM_PASSWORD_LENGTH
+                    || user.getPassword() == null) {
+            throw new ValidationException("PASSWORD " + user.getPassword() + "is to short");
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:21]: Message says that password is too short, but if password is null you don't provide correct exception message.

**Review by Elena-Bruyako — APPROVED** [review]
Approve, but pay attention to the comment
